libraries/chain/wasm_alloc_pool.cpp

```diff
@@ -0,0 +1,92 @@
+#include <eosio/chain/wasm_alloc_pool.hpp>
+#include <fc/log/logger.hpp>
+
+namespace eosio::chain {
+
+// only called on main thread
+wasm_alloc_pool::wasm_alloc_pool()
+   : num_threads(1)
+   , max_call_depth(1)
+{
+   // create 1 wasm allocator for the main thread
+   stack = std::make_unique<boost::lockfree::stack<vm::wasm_allocator*>>(1);
+   stack->push(new vm::wasm_allocator);
+}
+
+wasm_alloc_pool::~wasm_alloc_pool() {
+   vm::wasm_allocator* alloc;
+   while (stack->pop(alloc)) {
+      delete alloc;
+   }
+}
+
+// called on any threads
+vm::wasm_allocator* wasm_alloc_pool::acquire() {
+   // Each thread can use at most `max_sync_call_depth` wasm allocators
+   // The stack would never be empty for a new acquire request
+   assert(!stack->empty());
+
+   vm::wasm_allocator* alloc;
+   stack->pop(alloc);
+
+   assert(alloc);
+   return alloc;
+}
+
+// called on any threads
+void wasm_alloc_pool::release(vm::wasm_allocator* alloc) {
+   stack->push(alloc);
+}
+
+// called on main thread from producer_plugin startup number of read-only threads is determined
+void wasm_alloc_pool::set_threads(uint32_t new_num_thread) {
+   if (new_num_thread <= num_threads) {
+      // For simplicity, we don't shrink the pool
+      return;
+   }
+
+   resize(new_num_thread, max_call_depth);
+}
+
+// called on main thread from sync_call protocol feature activation or set_packed_parameters
+void wasm_alloc_pool::set_max_call_depth(uint32_t new_depth) {
+   if (new_depth <= max_call_depth) {
+      // For simplicity, we don't shrink the pool
+      return;
+   }
+
+   resize(num_threads, new_depth);
+}
+
+void wasm_alloc_pool::resize(uint32_t new_num_thread, uint32_t new_depth) {
+   auto old_pool_size  = num_threads * max_call_depth; // take a note of old size
+
+   // save allocators not being in use into a temporary place
+   std::vector<vm::wasm_allocator*> temp;
+   vm::wasm_allocator* alloc;
+   while (stack->pop(alloc)) {
+      temp.push_back(alloc);
+   }
+
+   // update with new values
+   num_threads         = new_num_thread;
+   max_call_depth      = new_depth;
+   auto new_pool_size  = num_threads * max_call_depth;
+
+   // recreate the internal stack with the new capacity
+   stack = std::make_unique<boost::lockfree::stack<vm::wasm_allocator*>>(new_pool_size);
```

Comment:
Why recreate the stack? You are using a variable-sized stack, seems like you can just add to it.

Also as is, `resize` can't be called concurrently with `acquire()` or `release()`. That seems fine, as those should only be called when the main-thread is idle. If you use the same stack and make `num_threads` and `max_call_depth` atomic you could make it safe to be called concurrently. Since we don't have that use-case, seems fine to just add a comment about the restriction.

Reply:
By doing `make_unique` you're effectively adding a potential lock anyways since the system allocator may lock. afaict there is no effective difference in lockiness by doing the `make_unique` over letting the queue allocate more memory as needed since that would use the system allocator too. So i'm not really seeing a benefit in recreating it.

Reply:
It helps with lockfree because it avoids `new`. But your use of `new` negates that. I think you should not add `boost::lockfree::fixed_sized<true>` and just let it dynamically grow without the re-create.

Reply:
The Boost manual says

> boost::lockfree::fixed_sized<>, defaults to boost::lockfree::fixed_sized<false>
> Can be used to completely disable dynamic memory allocations during push in order to ensure lockfree behavior.
> If the data structure is configured as fixed-sized, the internal nodes are stored inside an array and they are addressed by array indexing. This limits the possible size of the stack to the number of elements that can be addressed by the index type (usually 2**16-2), but on platforms that lack double-width compare-and-exchange instructions, this is the best way to achieve lock-freedom.

Reply:
afaik As long as you `reserve()` the appropriate amount of space, and if you use `bounded_push()` the push will always be lock free. I think even `push()` will still be lockfree as long as you're still not exceeding what was reserved.

Reply:
Thank you both!

I removed the recreation. `resize()` is much simpler.
https://github.com/AntelopeIO/spring/pull/1346/commits/5dffde4ff4a811e7988510217f3d2ee43653e472